File: src/agents/ali/2.cc
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

struct Interval {
  int r;
  int l;
  long long w;
};

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int t;
  cin >> t;
  while (t--) {
    int n, m;
    cin >> n >> m;

    vector<long long> prefix(n + 1, 0);
    for (int i = 1; i <= n; ++i) {
      long long x;
      cin >> x;
      prefix[i] = prefix[i - 1] + x;
    }

    vector<Interval> intervals;
    intervals.reserve(m);
    for (int i = 0; i < m; ++i) {
      int l, r;
      cin >> l >> r;
      long long w = prefix[r] - prefix[l - 1];
      intervals.push_back({r, l, w});
    }

    sort(intervals.begin(), intervals.end(),
         [](const Interval& a, const Interval& b) {
           if (a.r != b.r) {
             return a.r < b.r;
           }
           return a.l < b.l;
         });

    vector<int> ends(m);
    for (int i = 0; i < m; ++i) {
      ends[i] = intervals[i].r;
    }

    vector<long long> dp(m + 1, 0);
    for (int i = 1; i <= m; ++i) {
      const auto& cur = intervals[i - 1];
      int prev = lower_bound(ends.begin(), ends.end(), cur.l) - ends.begin();
      dp[i] = max(dp[i - 1], dp[prev] + cur.w);
    }

    cout << dp[m] << '\n';
  }

  return 0;
}

File: src/agents/ali/2.ts
import fs from "fs";

interface Interval {
  r: number;
  l: number;
  w: number;
}

function lowerBound(arr: number[], target: number): number {
  let left = 0;
  let right = arr.length;
  while (left < right) {
    const mid = (left + right) >> 1;
    if (arr[mid] < target) {
      left = mid + 1;
    } else {
      right = mid;
    }
  }
  return left;
}

function solve(input: string): string {
  const data = input.trim().split(/\s+/).map(Number);
  if (data.length === 0 || Number.isNaN(data[0])) {
    return "";
  }

  let idx = 0;
  const t = data[idx++];
  const answers: string[] = [];

  for (let tc = 0; tc < t; tc++) {
    const n = data[idx++];
    const m = data[idx++];

    const prefix = new Array<number>(n + 1).fill(0);
    for (let i = 1; i <= n; i++) {
      prefix[i] = prefix[i - 1] + data[idx++];
    }

    const intervals: Interval[] = [];
    for (let i = 0; i < m; i++) {
      const l = data[idx++];
      const r = data[idx++];
      const w = prefix[r] - prefix[l - 1];
      intervals.push({ r, l, w });
    }

    intervals.sort((a, b) => (a.r !== b.r ? a.r - b.r : a.l - b.l));

    const ends = intervals.map((item) => item.r);
    const dp = new Array<number>(m + 1).fill(0);

    for (let i = 1; i <= m; i++) {
      const cur = intervals[i - 1];
      const prev = lowerBound(ends, cur.l);
      dp[i] = Math.max(dp[i - 1], dp[prev] + cur.w);
    }

    answers.push(String(dp[m]));
  }

  return answers.join("\n");
}

const input = fs.readFileSync(0, "utf8");
const output = solve(input);
if (output.length > 0) {
  process.stdout.write(output);
}
